### PR TITLE
[TSSDKv2][2/n] add Multi Ed25519 classes

### PR DESCRIPTION
**Pull request by 0xjinn**
### Description

Note: This is stack on top of #10157

This PR adds the following classes

1. MultiEd25519PublicKey
2. MultiEd25519Signature

This version follows the same design as Python SDK

### Test Plan

jest test

### CODE REVIEW

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+   * @see {@link
+   * https://aptos.dev/integration/creating-a-signed-transaction/#multisignature-transactions | Creating a Signed Transaction}
+   *
+   * @param args.signatures A list of signatures
+   * @param args.bitmap 4 bytes, at most 32 signatures are supported. If Nth bit value is `1`, the Nth
+   * signature should be provided in `signatures`. Bits are read from left to right
+   */
+  constructor(args: { signatures: Signature[]; bitmap: Uint8Array }) {
+    const { signatures, bitmap } = args;
+    if (bitmap.length !== MultiSignature.BITMAP_LEN) {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:109]: do we also want to check args.signatures length?

File: ecosystem/typescript/sdk_v2/tests/unit/multi_ed25519.test.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+import { Deserializer } from "../../src/bcs/deserializer";
+import { Serializer } from "../../src/bcs/serializer";
+import { Hex } from "../../src/core/hex";
+import { PublicKey, Signature } from "../../src/crypto/ed25519";
+import { MultiPublicKey, MultiSignature } from "../../src/crypto/multi_ed25519";
+import { multiEd25519PkTestObject, multiEd25519SigTestObject } from "./helper";
+
+describe("MultiPublicKey", () => {
+  it("should serializes to bytes correctly", async () => {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/tests/unit/multi_ed25519.test.ts:12]: should we also test `toUint8Array()`?

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  // Minimum number of public keys required
+  static readonly MIN_KEYS = 2;
+
+  // Minimum number of threshold supported
+  static readonly MIN_THRESHOLD = 1;
+
+  // List of Ed25519 public keys for this MultiEd25519PublicKey
+  public readonly publicKeys: PublicKey[];
+
+  // At least "threshold" signatures must be valid for the number of public keys specified
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:21]: can we be more clear with this comment? for me `At least` is a bit confusing given we have `MIN_THRESHOLD`

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+   * Here's an example of valid `bits`
+   * ```
+   * [0, 2, 31]
+   * ```
+   * `[0, 2, 31]` means the 1st, 3rd and 32nd bits should be set in the bitmap.
+   * The result bitmap should be 0b1010000000000000000000000000001
+   *
+   * @returns bitmap that is 32bit long
+   */
+  static createBitmap(bits: number[]): Uint8Array {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:145]: should it be
```suggestion
  static createBitmap(args:{bits: number[]}): Uint8Array {
```

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    const bytes = deserializer.deserializeBytes();
+    const bitmap = bytes.subarray(bytes.length - 4);
+
+    const sigs: Signature[] = [];
+
+    for (let i = 0; i < bytes.length - bitmap.length; i += Signature.LENGTH) {
+      const begin = i;
+      sigs.push(new Signature({ data: bytes.subarray(begin, begin + Signature.LENGTH) }));
+    }
+    return new MultiSignature({ signatures: sigs, bitmap });
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:191]: nit - change `sigs` to `signatures` and then can
```suggestion
    return new MultiSignature({ signatures, bitmap });
```

**Review by 0xmaayan — APPROVED** [review]
LGTM

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts (hunk 1 above)
Comment (0xjinn) [reply ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:109]: Good catch!

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts (hunk 3 above)
Comment (0xjinn) [reply ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:21]: Yep, will be updating this to `The minimum number of valid signatures required, for the number of public keys specified`

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts (hunk 4 above)
Comment (0xjinn) [reply ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:145]: Curious, do we want to make everything obj for argument, including Serialize and Deserialize method as well?

**Review by davidiw — COMMENTED** [review]
why don't we have functions like verify in here? or actually store the public keys as Ed25519Keys or something?

**Comment by 0xjinn** [comment]
> why don't we have functions like verify in here?

@davidiw Wanted to make sure. Do you mean `verify` in MultiPublicKey or PublicKey?

> or actually store the public keys as Ed25519Keys or something?

The MultiPublicKey actually stores an array of `PublicKey`. Hmm am I missing something here?

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    // Check if duplicates exist in bits
+    const dupCheckSet = new Set();
+
+    bits.forEach((bit: number) => {
+      if (bit >= MultiSignature.MAX_SIGNATURES_SUPPORTED) {
+        throw new Error(`Invalid bit value ${bit}.`);
+      }
+
+      if (dupCheckSet.has(bit)) {
+        throw new Error("Duplicated bits detected.");
```
Comment (gregnazario) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:165]: `Duplicate bits detected`

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+    // The decimal value of 0b10000000 is 128.
+    const firstBitInByte = 128;
+    const bitmap = new Uint8Array([0, 0, 0, 0]);
+
+    // Check if duplicates exist in bits
+    const dupCheckSet = new Set();
+
+    bits.forEach((bit: number) => {
+      if (bit >= MultiSignature.MAX_SIGNATURES_SUPPORTED) {
+        throw new Error(`Invalid bit value ${bit}.`);
```
Comment (gregnazario) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:161]: `Cannot have a signature larger than ${MultiSignature.MAX_SIGNATURES_SUPPORTED - 1}`

**Review by gregnazario — COMMENTED** [review]
Only comment I really have is, we probably need support in the Account object to tell if it will properly work.  This works in a bubble here, but it needs full compatibility with the Rust code.

I saw some code in @davidiw's PRs that led me to believe the MultiEd25519 stuff is serialized in a couple different ways.

File: ecosystem/typescript/sdk_v2/tests/unit/multi_ed25519.test.ts [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+      ],
+      threshold: 2,
+    });
+
+    const expected = new Uint8Array([
+      185, 198, 238, 22, 48, 239, 62, 113, 17, 68, 166, 72, 219, 6, 187, 178, 40, 79, 114, 116, 207, 190, 229, 63, 252,
+      238, 80, 60, 193, 164, 146, 0, 174, 243, 244, 164, 184, 236, 161, 223, 195, 67, 54, 27, 248, 228, 54, 189, 66,
+      222, 146, 89, 192, 75, 131, 20, 235, 142, 32, 84, 221, 110, 130, 171, 138, 87, 98, 226, 26, 193, 205, 179, 135, 4,
+      66, 199, 123, 76, 58, 245, 140, 124, 237, 184, 119, 157, 2, 112, 230, 212, 241, 226, 247, 54, 125, 116, 2,
+    ]);
```
[the remark this answers is not in the document]
Comment (0xjinn) [reply ecosystem/typescript/sdk_v2/tests/unit/multi_ed25519.test.ts:31]: Yep, I think there are other tests that are using Uint8Array as well, will clean all of the tests up afterward.

Issue created - #10286

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+  serialize(serializer: Serializer): void {
+    serializer.serializeBytes(this.toUint8Array());
+  }
+
+  // TODO: Update this in interface to be static, then remove this method
+  deserialize(deserializer: Deserializer): Signature {
+    throw new Error("Method not implemented.");
+  }
+
+  static deserialize(deserializer: Deserializer): MultiEd25519Signature {
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:210]: remove?

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+export class MultiEd25519Signature extends Signature {
+  // Maximum number of signatures supported
+  static MAX_SIGNATURES_SUPPORTED = 32;
+
+  // Bitmap length
+  static BITMAP_LEN: number = 4;
+
+  public readonly signatures: Ed25519Signature[];
+
+  public readonly bitmap: Uint8Array;
```
Comment (0xmaayan) [inline ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:110]: lets make sure we have comments on class variables and methods explaining each

File: ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts (hunk 9 above)
Comment (0xjinn) [reply ecosystem/typescript/sdk_v2/src/crypto/multi_ed25519.ts:210]: I believe we still need this deserialize static method. The one above this one will be removed once the Deserializable interface is update.

**Review by gregnazario — APPROVED** [review]

**Pull request merged**